### ISSUE TITLE
Rejecting costly or malicious queries

### ISSUE THREAD

**Issue by herojan** [graphql-java/graphql-java#291]
Is there currently a way to determine how costly a query will be to run? I have been looking at the code and I think there isn't, correct me and close this issue if I'm wrong!

It would be important to add some protections against this kind of thing. When thinking of ways to do this I found that I settled on a way that [sangria](http://sangria-graphql.org/learn/#protection-against-malicious-queries) is already doing. i.e. assign a complexity score to each field and add up the scores, rejecting the query if it's too complex.

With this in mind I'd propose the following: 

1. Add an optional cost/complexity function to the GraphQLFieldDefinition and its inner Builder class. The interface would be something like:

 ```java
public interface ComplexityFunction {
    Double calculate(ValidationContext validationContext, List<GraphQLArgument> args, Double childScore);
}
```
with java 8 usage like
```java
newFieldDefinition()
.name("field")
.dataFetcher(env -> "result")
.complexity((validationContext, fieldDef, childScore) -> 2.0 * getArgument("size", args) * childScore)
```

2. Add a way to specify the complexity limit on queries to run. This will likely take the form of adding a List of validations as an argument to the `graphql.execute` method. It would be nice to be able to use the RulesVisitor and another validation for this but I'm not sure it can be done that way yet.

What do you think? I'm willing to implement this feature myself, so I'd like to know if this is wanted and if this is a good approach.

**Comment by dminkovsky** [comment]
Thank you for reporting @herojan. 

@timtebeek mentions this might also be related https://github.com/oembedler/spring-graphql-common#protection-against-malicious-queries

**Comment by herojan** [comment]
@dminkovsky Ah, I wasn't looking for a solution to a problem I was having, I already have support for this in my own project. I was making this issue to discuss whether or not this feature should be added to graphql-java. Actually adding it isn't hard.

I had seen that Spring project before but I didn't know it does this. It looks like they based their implementation on Sangria's too, since the documentation structure and headings are the same. The expression language is quite nice.

The idea was to add this to GraphQLFieldDefinition and its Builder class, in the same way as the dataFetcher is added. I modified the example above for it to be a bit clearer.

I do believe it would make sense here because it is a simple pre-execution analysis similar to the validation rules, and because no graphql interface could ever be exposed without supporting this kind of thing.

**Comment by bbakerman** [comment]
I had a got with this PR #299  to being constraints.  It only does depth for the moment but I think it can be extended to be the place for "field complexity"

**Comment by Macroz** [comment]
See also issue https://github.com/graphql-java/graphql-java/issues/75

**Comment by andimarek** [comment]
I would like to see some kind of solution for malicious queries in the next version: the question what to do about it comes up all the time and I think graphql-java needs to have some kind of answer to that.

**Comment by kaqqao** [comment]
I'm almost done implementing a pretty generic support for this in graphql-spqr. Here's how I've done it in broad terms, maybe it gives you some ideas.

* Implemented an AST walker, that basically does what `FieldCollector` does, but for the entire tree (not just one level, like `FieldCollector`)
* The walker accepts visitor-style callbacks that it will trigger before and after each node is processed, giving the visitor access to the current `Field`, `GraphQLFieldDefinition`, argument values, and sub-selection
* Because sub-selections are provided, the tree will be traversed in reverse (leaf nodes will be visited first)
* Because the tree-walking is generic, a visitor likely has to be stateful
* Visitors I already have or intend to implement:
  * Static complexity analysis, in the same style as [Sangria](https://github.com/sangria-graphql/sangria) or [Absinthe](http://absinthe-graphql.org/). This visitor maintains an accumulated complexity score that it increments as each node is visited. Once a threshold is reached, it throws, and the entire query resolution fails.
  * Dynamic complexity analysis. The visitor builds a map of complexity _computations_ for each node, and stores it in the context. Later on, as the query is being executed, just before a leaf (scalar or enum) fetcher runs, it completes the computation for the leaf node. This triggers re-evaluation of the parent node's complexity score, in turn triggering its parent, and so forth up to the root. Basically, when ever a leaf node is about to be resolved, complexity scores of everything above it are recalculated. Once a threshold os reached at any level, an exception is thrown and the execution is aborted. Currently, this has no benefits over static analysis as graphql-java provides no way to return the current partial result when instrumentation throws an error, but @andimarek mentioned this is about to change, so the benefit of this strategy is that it will return partial results for overly complex queries instead of outright rejecting them.
  * Execution time analysis. This visitor will increment cumulative execution time after each node is resolved. Once the global threshold is reached, it throws.

Whatever ends up in graphql-java itself, I'd argue it's of utmost importance to make it extensible/customizable, so that different behaviors can be plugged in easily. Then projects and libraries can provide concrete implementations and/or extensions.

**Comment by ieugen** [comment]
FYI, Github v4 api is GraphQL based and it's interesting to see how they enforce limits [1]. 

They have a node limit and a rate limit: 

> Clients must supply a first or last argument on any connection.
  Values of first and last must be within 1-100.
  Individual calls cannot request more than 11,250 total nodes.


[1] https://developer.github.com/v4/guides/resource-limitations/

**Comment by kaqqao** [comment]
Just in case someone cares, my implementation is ready. It's in [this](https://github.com/leangen/graphql-spqr/tree/master/src/main/java/io/leangen/graphql/execution/instrumentation) package, the main class being [ComplexityAnalyzer](https://github.com/leangen/graphql-spqr/blob/master/src/main/java/io/leangen/graphql/execution/instrumentation/ComplexityAnalyzer.java).

Functionally, it replicates Sangria's approach, It uses JavaScript expressions for the dynamic formulas (so no dependency on any expression language), but this is easily customizable, so other languages (SpEL, OGNL, UEL etc), calculation methods or sources of info (think of a DB with live performance analytics) are pluggable.

**Comment by andimarek** [comment]
this is now solved with #655: based on query depth and/or query complexity you can now prevent the execution of certain queries.
